mmvb_backend/benchmarking_sessions/api/views.py

```diff
@@ -0,0 +1,124 @@
+from django.shortcuts import get_object_or_404
+from rest_framework import status
+from rest_framework.decorators import action
+from rest_framework.response import Response
+from rest_framework.schemas.openapi import AutoSchema
+from rest_framework.viewsets import ModelViewSet
+
+from benchmarking_sessions.api.serializers import (
+    BenchmarkingSessionResultsSerializer,
+    BenchmarkingSessionSerializer,
+)
+from benchmarking_sessions.models import (
+    BenchmarkingSession,
+    BenchmarkingStepStatus,
+)
+from benchmarking_sessions.tasks import run_benchmark
+
+
+# todo: properly document endpoints
+class BenchmarkingSessionViewSet(ModelViewSet):
+    schema = AutoSchema(tags=["BenchmarkingSessions",])
+    serializer_class = BenchmarkingSessionSerializer
+
+    def get_queryset(self):
+        return BenchmarkingSession.objects.all()
+
+    # todo: generate 404
+    @action(methods=["post"], detail=True, url_path="run")
+    def run_benchmark(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
+            BenchmarkingSession, id=kwargs["pk"]
+        )
+        task = run_benchmark.delay(benchmarking_session.id)
+        benchmarking_session.task_id = task.id
+        benchmarking_session.save()
+        return Response(
+            f"/benchmarking-sessions/{benchmarking_session.id}/status",
+            status=status.HTTP_202_ACCEPTED,
+        )
+
+    # todo: document response structure (OpenAPI)
+    @action(methods=["get"], detail=True, url_path="status")
+    def benchmark_status(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
+            BenchmarkingSession, id=kwargs["pk"]
+        )
+        if benchmarking_session.status != BenchmarkingSession.Status.RUNNING:
+            return Response(
+                {"status": benchmarking_session.status},
+                status=status.HTTP_200_OK,
+            )
+
+        result = run_benchmark.AsyncResult(benchmarking_session.task_id)
+        if result.status == "PENDING" or result.info is None:
+            # this is an in-between state where the celery task status is not yet reflected in the main DB,
+            # so either the task is about to begin or just finished -- in either case it's easiest to pretend it is
+            # running
+            return Response(
+                {"status": BenchmarkingSession.Status.INTERMEDIATE},
+                status=status.HTTP_200_OK,
+            )
+
+        for case_responses in result.info["responses"]:
```

Comment:
This is just cleaning up the results, right?
Might be worthy to have a separate reusable function to do the cleaning here.

Reply:
sure – where would be the django place to put such a function?
(it's not really "cleaning up" – it's reducing the payload for the status update. there is no interest in the actual responses at this stage, only whether the system received a response or if not, what kind of error)

Reply:
I guess this is not related to a "django pattern" or if there is one I'm not aware. I would check:
  - Would this "reducing" function be more closely related to api views? If so, I would create a `utils.py` inside `benchmarking_sessions/api` and have it implemented there.
  - Or does it have a possible broader usage within this application? If so, I would create a `utils.py` inside `benchmarking_sessions`.

Having it more self-contained like this would help improve reusability, code readability, and also make it easier to implement more self-contained unit tests.